lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -863,6 +866,14 @@ struct FIRRTLLowering : public LowerFIRRTLToRTLBase<FIRRTLLowering>,
   /// key should have a FIRRTL type, the result will have an RTL dialect type.
   DenseMap<Value, Value> valueMapping;
 
+  /// Each register lowered needs to have connects placed in the always block,
+  /// and the else reset block.  This tracks the block for each lowered
+  /// register.
+  DenseMap<std::pair<Value, Value>, std::pair<Block *, Block *>> regBlocks;
+
+  /// Each value used as a clock generates one always block.
+  DenseMap<Value, Block *> clockBlocks;
```

Comment:
The general structure of LowerToRTL is to emit new always/if/initial etc blocks, and have a cleanup pass (yet to be written) that merges them together, folding conditions etc.

Does this structure work this this, and eliminate the need for the maps?  It looks like these exist just to make prettier IR.

Reply:
So it turns out no.  Registers of different type (asyncreset, syncreset, noreset) need to be in separate always blocks.
http://www.sunburst-design.com/papers/CummingsSNUG2003Boston_Resets.pdf

Reply:
What isn't clear is whether tools would be ok with this when mixing followers and reset registers:
```
always @posedge clk begin
  followerreg <=  val
  if (reset) begin
     syncreg <= val2
   end else begin
     syncreg <= val3
  end
end
```

Reply:
Probably not. There may also be additional restrictions around asynchronous resets like [the Formality requirement of a single statement in an asynchronous reset always block](https://github.com/chipsalliance/firrtl/blob/15013df6f6ac2dafeb35d7ed15cf95c7ac8a5bef/src/main/scala/firrtl/backends/verilog/VerilogEmitter.scala#L573).

I'd be in favor of one block for each `(Clock, Option<ResetType>)` tuple and not merge beyond that.

Reply:
Ok, I wasn't aware of that.  What does that mean for the semantics of always?  How do we reason about what can and what can't be merged?

Reply:
Merging after we're at the level of `always` may be tricky as we've lost type information about what is a reset. E.g., it's unclear in the above example that `reset` is actually a synchronous reset and should block the merge.

If this is coming from FIRRTL, the information is available and merging can happen, but this conflicts with the approach of lowering to `always` and then merging. Maybe this motivates another IR construct that represents registers or a "higher level" always block that takes an `(Clock, Option<ResetType>)` input? (This pushes towards an IR construct for registers like the "sequential" dialect discussion.)

Reply:
I think much of this goes away when the RTL dialect has a register abstraction...

Some of the complication in merging will come from having reset and !reset floating around as conditions.

But for now, I've updated to go back to the always blocks per statement.

It seems we may have to tread always blocks as having a list of sensitvities and a reset signal so we can keep things sorted out.

Reply:
> It seems we may have to tread always blocks as having a list of sensitvities and a reset signal so we can keep things sorted out.

Yeah I think this is the right path.  We can have two forms of always blocks:  

1) the current syntactic one which is basically a "don't touch" for the optimizer because the frontend didn't provide metadata that makes it safe to do various transformations (I still don't grok exactly why, is this pattern matching on the part of downstream tools? ELI5 plz :-)

2) the current one with a list of additional metadata (that the frontend can optionally provide) which, when present, allows safe merging and transformation.  FIRRTL lowering can provide this if/since we already have it.

I think that this is going to be non-obvious, so splitting it out into its own discussion (probably on the forum) would make sense.  In the meantime, thank you for doing the "technically wrong but consistent with the current direction" approach @darthscsi